rfc-123-github-actions-ci.md

```diff
@@ -50,6 +50,20 @@ This would be a gradual process, because we have many repos, but this decision w
 - Another system to learn, as well as the replatforming work already ongoing.
 - Duplication of build parameters and steps between all our apps, as they’re all Rails apps.
 
+### Findings from some initial explorations into using GitHub Actions
+
+#### `push` vs `pull_request` build triggers
+
+There has been some [chatter](https://github.com/alphagov/govuk-developer-docs/pull/2497#discussion_r415572147) about if we want `push` or `pull_request` triggers for tests to run, or both. The GOV.UK Developer Tools group has done some investigation into the differences between this, as there was some concern about if we wanted long-running builds (in the case of apps with large test suites, like whitehall) to run twice when not needed. There is not currently any way to differentiate and say "run on PRs from forks" and "run on push events to main repo branches". We'd have really liked a `pull_request_only_from_forks` setting, but that doesn't exist. It's worth noting that this isn't a problem _only_ with GitHub Actions: [Travis](https://github.com/travis-ci/travis-ci/issues/1147) has had the same pain points for years, and we currently do double builds for the apps we use Travis for.
+
+Some solutions to this include:
+
+- Running `on: [pull_request, push]`. This means that internal devs get feedback on their commits when they push them, for example if they've forgotten to run RuboCop locally. And also so that external contributors PRs get CI feedback on their changes. It does mean that for long-running test suites, we could waste compute power, time and the planet's precious natural resources running things twice that don't need to be run twice.
+
+- Running `on: push`. Actions don't run [by default on forked PRs](https://github.com/alphagov/govuk-docker/pull/337) - it's a security feature that GitHub has implemented. And [when a PR is raised from a fork](https://github.com/alphagov/govuk-docker/pull/337) and the receiving repo only has a push trigger, the tests don't run. So we'd be back to the annoying state that is [our current process for forked PRs][merge-a-pull-request].
```

Comment:
Should be flagged here that this doesn't merge with master so we only know the branch passes and not if the integrated branch passes.

Reply:
GitHub does have [an option in the branch protection settings](https://help.github.com/en/github/administering-a-repository/types-of-required-status-checks) to enforce that branches are up to date with `master` before merging. But [we don't currently enforce it](https://github.com/alphagov/govuk-saas-config/blob/master/github/lib/configure_repo.rb#L95).

Reply:
I don't think many people rebase before merging, so switching that on would be a big disruptive change for a lot of people.

Reply:
I don't think it would be particularly disruptive actually. People on long-running branches regularly rebase against master, in my experience - and for people doing small tweaks, GitHub offers a "Rebase to master" button right on the PR itself.

The only disruption is waiting for the tests to pass again, especially if there's a flaky publishing-e2e-tests downstream too. I guess it might encourage us to make our test suites faster and more reliable.

Reply:
I'm definitely an advocate of rebasing before pushing, just to make sure that everything runs fine against the latest master code. It can be disruptive in a fast-moving repo though as Chris points out above if in the time tests take to run, more commits are pushed.